crates/sui-types/src/gas.rs

```diff
@@ -2,103 +2,139 @@
 
-use crate::error::{UserInputError, UserInputResult};
-use crate::messages::TransactionEffects;
 use crate::{
-    error::{ExecutionError, ExecutionErrorKind},
+    error::{ExecutionError, UserInputError, UserInputResult},
     gas_coin::GasCoin,
-    messages::TransactionEffectsAPI,
-    object::{Object, Owner},
+    gas_model::gas_v1::{self, SuiCostTable as SuiCostTableV0, SuiGasStatus as SuiGasStatusV0},
+    messages::{TransactionEffects, TransactionEffectsAPI},
+    object::Object,
 };
+use enum_dispatch::enum_dispatch;
 use itertools::MultiUnzip;
-use move_core_types::{
-    gas_algebra::{GasQuantity, InternalGas, InternalGasPerByte, NumBytes, UnitDiv},
-    vm_status::StatusCode,
-};
-use once_cell::sync::Lazy;
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
-use std::ops::AddAssign;
-use std::{
-    convert::TryFrom,
-    ops::{Add, Deref, Mul},
-};
-use sui_cost_tables::{
-    bytecode_tables::{GasStatus, INITIAL_COST_SCHEDULE},
-    units_types::GasUnit,
-};
-use sui_protocol_config::*;
-
-macro_rules! ok_or_gas_balance_error {
-    ($balance:expr, $required:expr) => {
-        if $balance < $required {
-            Err(UserInputError::GasBalanceTooLow {
-                gas_balance: $balance,
-                needed_gas_amount: $required,
-            })
-        } else {
-            Ok(())
-        }
-    };
-}
+use sui_cost_tables::bytecode_tables::GasStatus;
+use sui_protocol_config::ProtocolConfig;
 
 sui_macros::checked_arithmetic! {
 
-// A bucket defines a range of units that will be priced the same.
-// A cost for the bucket is defined to make the step function non linear.
-#[allow(dead_code)]
-struct ComputationBucket {
-    min: u64,
-    max: u64,
-    cost: u64,
+#[enum_dispatch]
+pub trait SuiGasStatusAPI<'a> {
+    fn is_unmetered(&self) -> bool;
+    fn bucketize_computation(&mut self) -> Result<(), ExecutionError>;
+    fn summary(&self) -> GasCostSummary;
+    fn gas_budget(&self) -> u64;
+    fn storage_gas_units(&self) -> u64;
+    fn storage_rebate(&self) -> u64;
+    fn gas_used(&self) -> u64;
+    fn reset_storage_cost_and_rebate(&mut self);
+    fn charge_storage_read(&mut self, size: usize) -> Result<(), ExecutionError>;
+    fn charge_storage_mutation(
+        &mut self,
+        new_size: usize,
+        storage_rebate: u64,
+    ) -> Result<u64, ExecutionError>;
+    fn charge_publish_package(&mut self, size: usize) -> Result<(), ExecutionError>;
+    fn move_gas_status(&mut self) -> &mut GasStatus<'a>;
+}
+
+#[enum_dispatch(SuiGasStatusAPI)]
+pub enum SuiGasStatus<'a> {
+    V0(SuiGasStatusV0<'a>),
 }
 
-impl ComputationBucket {
-    fn new(min: u64, max: u64, cost: u64) -> Self {
-        ComputationBucket { min, max, cost }
+impl<'a> SuiGasStatus<'a> {
+    pub fn new_with_budget(gas_budget: u64, gas_price: u64, config: &ProtocolConfig) -> Self {
+        match config.gas_model_version() {
+            1 => Self::V0(SuiGasStatusV0::new_with_budget(
```

Comment:
Perhaps `SuiGasStatus` and `SuiCostTable` should start at `V1` to avoid the potential confusion of `1 vs `V0`?

Reply:
they do now, wondering if I made the change on the other PR, but yeah I had the same issue when reviewing the code

Reply:
apparently I changed that in the new PR, sorry for that, but I think those PRs will be merged and splitting them was important for review so I guess it's all good